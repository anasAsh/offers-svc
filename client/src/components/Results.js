import React, { Component } from 'react';

const END_POINT = '/api/offers/?';

class Results extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: false,
      data: {}
    }
  }

  componentDidMount() {
    this.fetchData();
  }

  componentDidUpdate(prevProps) {
    if (JSON.stringify(prevProps.filters) !== JSON.stringify(this.props.filters)) {
      this.fetchData(this.props.filters);
    }
  }

  fetchData(filters) {
    const url = END_POINT + new URLSearchParams(filters).toString();
    this.setState({
      isLoading: true
    });

    fetch(url)
      .then(response => response.json())
      .then(data => this.setState({data, isLoading: false}))
      .catch(error => this.setState({error, isLoading: false}));
  }



  render() {
    const {error, data, isLoading} = this.state;
    if (error) {
      return `Error: ${JSON.stringify(error)}`;
    }
    if (isLoading) {
      return 'Loading...';
    }

    const { Hotel: deals = [] } = (data.offers || {});
    const items = deals.map((offer, index) => {
      return <li key={index}>
        <pre>{JSON.stringify(offer)}</pre>
      </li>
    });

    return (
      <div className="Results">
        Results
        <ul>
          {items}
        </ul>
      </div>


    );
  }
}

export default Results;
