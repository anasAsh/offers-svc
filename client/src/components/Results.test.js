import React from "react";
import { shallow, mount } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Results from "./Results";

configure({ adapter: new Adapter() });

describe('<Results /> component', () => {

  beforeAll(() => {

  });
  
  it('should match the snapshot', () => {
    const container = shallow(<Results />);
    expect(container.html()).toMatchSnapshot();
  });

  it('should have start loading', () => {
    const container = shallow(<Results />);
    expect(container.state('isLoading')).toEqual(true);
  });

  it('should fetch data', async(done) => {
    const fakeData = {
      anything: 1
    }
    const spy = jest.spyOn(global, "fetch").mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve(fakeData)
      })
    );
    const container = shallow(<Results />);
    setTimeout(() => {
      container.update();
      const state = container.instance().state;
      expect(state.data).toEqual(fakeData);
      expect(spy).toHaveBeenCalledWith('/api/offers/?');
      done();
    }, 0);
  });

  it('should fetch data with correct filters', async(done) => {
    const fakeData = {
      anything: 1
    }
    const spy = jest.spyOn(global, "fetch").mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve(fakeData)
      })
    );
    const filters = {filter1: 'someString'};
    const container = mount(<Results/>);
    container.setProps({filters})
    
    setTimeout(() => {
      container.update();
      const state = container.instance().state;
      expect(container.props().filters).toEqual(filters);
      expect(spy).toHaveBeenCalledWith('/api/offers/?filter1=someString');

      done();
    }, 0);
  });

  it('should render data', async(done) => {
    const fakeData = {
      offers: {Hotel: [{someData: 1}]}
    }
    jest.spyOn(global, "fetch").mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve(fakeData)
      })
    );
    const filters = {filter1: 'someString'};
    const container = mount(<Results/>);
    container.setProps({filters})
    
    setTimeout(() => {
      container.update();
      expect(container.html()).toMatchSnapshot();
      done();
    }, 0);
  });

  it('should handle error', async(done) => {
    jest.spyOn(global, "fetch").mockImplementation(() =>
      Promise.reject({})
    );
    const container = mount(<Results/>);
    
    setTimeout(() => {
      container.update();
      expect(container.html()).toMatchSnapshot();
      done();
    }, 0);
  });

});

