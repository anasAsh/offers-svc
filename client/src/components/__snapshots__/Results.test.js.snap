// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Results /> component should handle error 1`] = `"Error: {}"`;

exports[`<Results /> component should match the snapshot 1`] = `""`;

exports[`<Results /> component should render data 1`] = `"<div class=\\"Results\\">Results<ul><li><pre>{\\"someData\\":1}</pre></li></ul></div>"`;

exports[`<Results /> with no props should handle error 1`] = `"Error: {}"`;

exports[`<Results /> with no props should match the snapshot 1`] = `""`;

exports[`<Results /> with no props should render data 1`] = `"<div class=\\"Results\\">Results<ul><li><pre>{\\"someData\\":1}</pre></li></ul></div>"`;
